tests/integration_tests/sqla_models_tests.py

```diff
@@ -913,54 +913,102 @@ def test_extra_cache_keys_in_sql_expression(
 
 @pytest.mark.usefixtures("app_context")
 @pytest.mark.parametrize(
-    "sql_expression,expected_cache_keys,has_extra_cache_keys",
+    "sql_expression,expected_cache_keys,has_extra_cache_keys,item_type",
     [
-        ("'{{ current_username() }}'", ["abc"], True),
-        ("(user != 'abc')", [], False),
+        ("'{{ current_username() }}'", ["abc"], True, "columns"),
+        ("(user != 'abc')", [], False, "columns"),
+        ("{{ current_user_id() }}", [1], True, "metrics"),
+        ("COUNT(*)", [], False, "metrics"),
     ],
 )
 @patch("superset.jinja_context.get_user_id", return_value=1)
 @patch("superset.jinja_context.get_username", return_value="abc")
-@patch("superset.jinja_context.get_user_email", return_value="abc@test.com")
-def test_extra_cache_keys_in_columns(
-    mock_user_email,
+def test_extra_cache_keys_in_adhoc_metrics_and_columns(
     mock_username,
     mock_user_id,
     sql_expression,
     expected_cache_keys,
     has_extra_cache_keys,
+    item_type,
 ):
     table = SqlaTable(
         table_name="test_has_no_extra_cache_keys_table",
         sql="SELECT 'abc' as user",
         database=get_example_database(),
     )
+    base_type = "metrics" if item_type == "columns" else "columns"
```

Comment:
typo?
```suggestion
    base_type = "metrics" if item_type == "metrics" else "columns"
```

Reply:
this one is actually correct. When `item_type` is set to `columns`, we want to have an empty `metrics` list in the `base_query_obj`. When `item_type` is set to `metrics`, we want to have an empty `columns` list in it.

I agree this seems a little confusing too. Wondering if it would be better to just split this test in two.

Reply:
In that case I'd maybe introduce a new property `overrides` that's a `dict` that you can merge into the `query_obj`. In this case you'd then define it as
```
{
    "metrics": [
        ...
    ],
}
```

or which ever set of properties you want to override.

Reply:
Ok found a way to make this simpler. Going to commit a change here.

Reply:
yup, ended up doing something like that. I was having issues to do the override because the final dict was being constructed with `dict()` but removed that  and it worked.